src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.MoveTypeEditor.cs

```diff
@@ -0,0 +1,199 @@
+                var membersToRemove = GetMembersToRemove(root);
+                foreach (var member in membersToRemove)
+                {
+                    documentEditor.RemoveNode(member, SyntaxRemoveOptions.KeepNoTrivia);
+                }
+
+                var modifiedRoot = documentEditor.GetChangedRoot();
+
+                // add an empty document to solution, so that we'll have options from the right context.
+                var solutionWithNewDocument = projectToBeUpdated.Solution.AddDocument(newDocumentId, State.TargetFileNameCandidate, text: string.Empty);
+
+                // update the text for the new document
+                solutionWithNewDocument = solutionWithNewDocument.WithDocumentSyntaxRoot(newDocumentId, modifiedRoot, PreservationMode.PreserveIdentity);
+
+                // get the updated document, perform clean up like remove unused usings.
+                var newDocument = solutionWithNewDocument.GetDocument(newDocumentId);
+                newDocument = await CleanUpDocumentAsync(newDocument, CancellationToken).ConfigureAwait(false);
+
+                return newDocument.Project.Solution;
+            }
+
+            /// <summary>
+            /// update the original document and remove the type that was moved.
+            /// perform other fix ups as necessary.
+            /// </summary>
+            /// <param name="sourceDocument">original document</param>
+            /// <returns>an updated solution with the original document fixed up as appropriate.</returns>
+            private async Task<Solution> RemoveTypeFromSourceDocumentAsync(Document sourceDocument)
+            {
+                var documentEditor = await DocumentEditor.CreateAsync(sourceDocument, CancellationToken).ConfigureAwait(false);
+
+                AddPartialModifiersToTypeChain(documentEditor);
+                documentEditor.RemoveNode(State.TypeNode, SyntaxRemoveOptions.KeepNoTrivia);
+
+                var updatedDocument = documentEditor.GetChangedDocument();
+
+                updatedDocument = await CleanUpDocumentAsync(updatedDocument, CancellationToken).ConfigureAwait(false);
+
+                return updatedDocument.Project.Solution;
+            }
+
+            /// <summary>
+            /// Traverses the syntax tree of the forked document and
+            /// collects a list of nodes that are not being moved.
+            /// This list of nodes are then removed from the forked copy.
+            /// </summary>
+            /// <param name="root">root, of the syntax tree of forked document</param>
+            /// <returns>list of syntax nodes, to be removed from the forked copy.</returns>
+            private IEnumerable<SyntaxNode> GetMembersToRemove(SyntaxNode root)
+            {
+                var spine = new HashSet<SyntaxNode>();
+
+                // collect the parent chain of declarations to keep.
+                SyntaxNode node = State.TypeNode;
```

Comment:
can't var because of typing between `TTypeDecl` and `Syntaxnode` but will inline, thanks>